Source/UI/NewUI/PaymentForm/ViewModel/DefaultPaymentViewModel.swift

```diff
@@ -1,32 +1,83 @@
 import UIKit
 import Checkout
 
+private struct CardDetails {
+  var number: String?
+  var expiryDate: ExpiryDate?
+  var name: String?
+  var cvv: String?
+  var billingAddress: Address?
+  var phone: Phone?
+
+  init(number: String? = nil, expiryDate: ExpiryDate? = nil, name: String? = nil, cvv: String? = nil, billingAddress: Address? = nil, phone: Phone? = nil) {
+    self.number = number
+    self.expiryDate = expiryDate
+    self.name = name
+    self.cvv = cvv
+    self.billingAddress = billingAddress
+    self.phone = phone
+  }
+
+  func getCard() -> Card? {
+    guard let number = number, let expiryDate = expiryDate else { return nil }
+    return Card(number: number,
+                expiryDate: expiryDate,
+                name: name,
+                cvv: cvv,
+                billingAddress: billingAddress,
+                phone: phone)
+  }
+}
+
 class DefaultPaymentViewModel: PaymentViewModel {
+  var updateLoading: (() -> Void)?
   var updateEditBillingSummaryView: (() -> Void)?
   var updateAddBillingDetailsView: (() -> Void)?
   var updateExpiryDateView: (() -> Void)?
   var updateCardNumberView: (() -> Void)?
-  var updateSecurityCodeView: (() -> Void)?
+  var updateSecurityCodeViewStyle: (() -> Void)?
+  var updatePayButtonView: (() -> Void)?
   var updateHeaderView: (() -> Void)?
+  var updateSecurityCodeViewScheme: ((Card.Scheme) -> Void)?
+  var shouldEnablePayButton: ((Bool) -> Void)?
+  var cardTokenRequested: ((Result<TokenDetails, TokenisationError.TokenRequest>) -> Void)?
   var supportedSchemes: [Card.Scheme]
   var cardValidator: CardValidator
   var logger: FramesEventLogging
+  var checkoutAPIService: CheckoutAPIProtocol
   var paymentFormStyle: PaymentFormStyle?
   var billingFormStyle: BillingFormStyle?
+  var currentScheme: Card.Scheme = .unknown
   var billingFormData: BillingForm? {
     didSet {
+      cardDetails.phone = billingFormData?.phone
+      cardDetails.billingAddress = billingFormData?.address
+      cardDetails.name = billingFormData?.name
       if isAddBillingSummaryNotUpdated() {
         updateBillingSummaryView()
       }
     }
   }
+  var isLoading: Bool = false {
+    didSet {
+      updateLoading?()
+    }
+  }
+
+  private var cardDetails: CardDetails = CardDetails() {
+    didSet {
+      shouldEnablePayButton?(cardDetails.expiryDate != nil && cardDetails.number != nil)
```

Comment:
Whilst @patrick-hoban-cko did say CVV may not be required, I think that was down to the developer to decide if they want it required or not? Shouldn't we use the `isMandatory` on the SecurityNumberStyle to decide if we include it or not?

Also, is this completion handler called every time we update a property within the object? I used to think this setter is called only when `setting the object`, which would probably be done at initialisation time?

Reply:
@alex-ioja-yang-cko, i think, this can be after isMandotry [ticket](https://checkout.atlassian.net/browse/PIMOB-1381?atlOrigin=eyJpIjoiZTJhMDJkY2RiNDlkNDU4OGE1MTY1NjlkN2NlM2QyMjAiLCJwIjoiaiJ9) 
Adding a [comment](https://checkout.atlassian.net/browse/PIMOB-1381?focusedCommentId=182223) to the ticket. it will be a small update for this logic.